In [ ]:
import numpy as Np
import pandas as Pd

import matplotlib.pyplot as Plt
%matplotlib inline

# Data Cleaning
Dataset may contain null/nan values and duplicates which can cost more time or can cause errors due to lack of information. So cleaning of data is an important task in data science.

In [ ]:
dataset = Pd.read_csv("/content/sample_data/movies.csv")

In [ ]:
#a = dataset.groupby(by="Station.State")
dataset

MOVIES         YEAR  \
0                           Blood Red Sky       (2021)   
1     Masters of the Universe: Revelation     (2021– )   
2                        The Walking Dead  (2010–2022)   
3                          Rick and Morty     (2013– )   
4                         Army of Thieves       (2021)   
...                                   ...          ...   
9994                       The Imperfects     (2021– )   
9995                               Arcane     (2021– )   
9996                    Heart of Invictus     (2022– )   
9997                       The Imperfects     (2021– )   
9998                       The Imperfects     (2021– )   

                                           GENRE  RATING  \
0         \nAction, Horror, Thriller                 6.1   
1     \nAnimation, Action, Adventure                 5.0   
2          \nDrama, Horror, Thriller                 8.2   
3     \nAnimation, Adventure, Comedy                 9.2   
4            \nAction, Crime, Horror                 NaN   
...                                          ...     ...   
9994     \nAdventure, Drama, Fantasy                 NaN   
9995  \nAnimation, Action, Adventure                 NaN   
9996            \nDocumentary, Sport                 NaN   
9997     \nAdventure, Drama, Fantasy                 NaN   
9998     \nAdventure, Drama, Fantasy                 NaN   

                                               ONE-LINE  \
0     \nA woman with a mysterious illness is forced ...   
1     \nThe war for Eternia begins again in what may...   
2     \nSheriff Deputy Rick Grimes wakes up from a c...   
3     \nAn animated series that follows the exploits...   
4     \nA prequel, set before the events of Army of ...   
...                                                 ...   
9994                                     \nAdd a Plot\n   
9995                                     \nAdd a Plot\n   
9996                                     \nAdd a Plot\n   
9997                                     \nAdd a Plot\n   
9998                                     \nAdd a Plot\n   

                                                  STARS    VOTES  RunTime  \
0     \n    Director:\nPeter Thorwarth\n| \n    Star...   21,062    121.0   
1     \n            \n    Stars:\nChris Wood, \nSara...   17,870     25.0   
2     \n            \n    Stars:\nAndrew Lincoln, \n...  885,805     44.0   
3     \n            \n    Stars:\nJustin Roiland, \n...  414,849     23.0   
4     \n    Director:\nMatthias Schweighöfer\n| \n  ...      NaN      NaN   
...                                                 ...      ...      ...   
9994  \n            \n    Stars:\nMorgan Taylor Camp...      NaN      NaN   
9995                                                 \n      NaN      NaN   
9996  \n    Director:\nOrlando von Einsiedel\n| \n  ...      NaN      NaN   
9997  \n    Director:\nJovanka Vuckovic\n| \n    Sta...      NaN      NaN   
9998  \n    Director:\nJovanka Vuckovic\n| \n    Sta...      NaN      NaN   

     Gross  
0      NaN  
1      NaN  
2      NaN  
3      NaN  
4      NaN  
...    ...  
9994   NaN  
9995   NaN  
9996   NaN  
9997   NaN  
9998   NaN  

[9999 rows x 9 columns]

### Removing Duplicates

In [ ]:
ds_red = dataset.groupby(by="MOVIES").agg("count")

dfsx = dataset.groupby(by="MOVIES")['MOVIES'].count()
print(dfsx)

print()

print("No of Distinct Rows = ", len(ds_red))
print("Difference = ",abs(len(ds_red) - len(dataset)) )

MOVIES
 13 Reasons Why              10
 1899                         8
 3Below: Tales of Arcadia    14
 50M2                         8
 7Seeds                      23
                             ..
Ófærð                         1
Órbita 9                      1
Ôkami to kôshinryô            1
Öldür Beni Sevgilim           1
ÜberWeihnachten               1
Name: MOVIES, Length: 6817, dtype: int64

No of Distinct Rows =  6817
Difference =  3182


The fact that grouping them based on the **name of movies** varies in length from the original dataset proves that there are around 3182 movie reviews which are **duplicates**

In [ ]:
reduced_ds = dataset.drop_duplicates("MOVIES",keep="first")

Notice that the duplicate reviews of movie "The Imperfects" is removed. similarly other duplicates will also be removed from dataset based on the given column "Movies".

You can see the great reduction in the count of total dataset, compared to the amount of data which are distinct.

In [ ]:
print(len(dataset))
print(len(reduced_ds))

9999
6817


### Handling missing data
We have two columns with major missing data : "Gross" and "Rating"

In [ ]:
#Shallow Copy
msh_ds = Pd.DataFrame(reduced_ds)
msh_ds = msh_ds.reset_index()

# Notice that Gross is not a necessary attribute, so we can drop it
msh_ds['Gross']

0       NaN
1       NaN
2       NaN
3       NaN
4       NaN
       ... 
6812    NaN
6813    NaN
6814    NaN
6815    NaN
6816    NaN
Name: Gross, Length: 6817, dtype: object

In [ ]:
msh_ds = msh_ds.drop('Gross',axis=1)
msh_ds

index                               MOVIES         YEAR  \
0         0                        Blood Red Sky       (2021)   
1         1  Masters of the Universe: Revelation     (2021– )   
2         2                     The Walking Dead  (2010–2022)   
3         3                       Rick and Morty     (2013– )   
4         4                      Army of Thieves       (2021)   
...     ...                                  ...          ...   
6812   9935                       The Imperfects     (2021– )   
6813   9964            AlRawabi School for Girls       (2021)   
6814   9993                            Totenfrau     (2022– )   
6815   9995                               Arcane     (2021– )   
6816   9996                    Heart of Invictus     (2022– )   

                                           GENRE  RATING  \
0         \nAction, Horror, Thriller                 6.1   
1     \nAnimation, Action, Adventure                 5.0   
2          \nDrama, Horror, Thriller                 8.2   
3     \nAnimation, Adventure, Comedy                 9.2   
4            \nAction, Crime, Horror                 NaN   
...                                          ...     ...   
6812     \nAdventure, Drama, Fantasy                 NaN   
6813                         \nDrama                 NaN   
6814               \nDrama, Thriller                 NaN   
6815  \nAnimation, Action, Adventure                 NaN   
6816            \nDocumentary, Sport                 NaN   

                                               ONE-LINE  \
0     \nA woman with a mysterious illness is forced ...   
1     \nThe war for Eternia begins again in what may...   
2     \nSheriff Deputy Rick Grimes wakes up from a c...   
3     \nAn animated series that follows the exploits...   
4     \nA prequel, set before the events of Army of ...   
...                                                 ...   
6812                                     \nAdd a Plot\n   
6813                                     \nAdd a Plot\n   
6814                                     \nAdd a Plot\n   
6815                                     \nAdd a Plot\n   
6816                                     \nAdd a Plot\n   

                                                  STARS    VOTES  RunTime  
0     \n    Director:\nPeter Thorwarth\n| \n    Star...   21,062    121.0  
1     \n            \n    Stars:\nChris Wood, \nSara...   17,870     25.0  
2     \n            \n    Stars:\nAndrew Lincoln, \n...  885,805     44.0  
3     \n            \n    Stars:\nJustin Roiland, \n...  414,849     23.0  
4     \n    Director:\nMatthias Schweighöfer\n| \n  ...      NaN      NaN  
...                                                 ...      ...      ...  
6812  \n            \n    Stars:\nMorgan Taylor Camp...      NaN      NaN  
6813  \n    Director:\nTima Shomali\n| \n    Stars:\...      NaN      NaN  
6814  \n    Director:\nNicolai Rohde\n| \n    Stars:...      NaN      NaN  
6815                                                 \n      NaN      NaN  
6816  \n    Director:\nOrlando von Einsiedel\n| \n  ...      NaN      NaN  

[6817 rows x 9 columns]

### Handling the Missing values in necessary rows
Time to replace the NaN's with default values

In [ ]:
msh_ds.describe()

index       RATING      RunTime
count  6817.000000  5815.000000  5305.000000
mean   3527.146252     6.635529    78.499529
std    2170.968436     1.210105    49.608718
min       0.000000     1.100000     1.000000
25%    1716.000000     5.900000    45.000000
50%    3439.000000     6.800000    82.000000
75%    5155.000000     7.500000   100.000000
max    9996.000000     9.400000   853.000000

In [ ]:
# Since the avg is 6.63
msh_ds['RATING'] = msh_ds['RATING'].fillna(value=6.0)
msh_ds

index                               MOVIES         YEAR  \
0         0                        Blood Red Sky       (2021)   
1         1  Masters of the Universe: Revelation     (2021– )   
2         2                     The Walking Dead  (2010–2022)   
3         3                       Rick and Morty     (2013– )   
4         4                      Army of Thieves       (2021)   
...     ...                                  ...          ...   
6812   9935                       The Imperfects     (2021– )   
6813   9964            AlRawabi School for Girls       (2021)   
6814   9993                            Totenfrau     (2022– )   
6815   9995                               Arcane     (2021– )   
6816   9996                    Heart of Invictus     (2022– )   

                                           GENRE  RATING  \
0         \nAction, Horror, Thriller                 6.1   
1     \nAnimation, Action, Adventure                 5.0   
2          \nDrama, Horror, Thriller                 8.2   
3     \nAnimation, Adventure, Comedy                 9.2   
4            \nAction, Crime, Horror                 5.0   
...                                          ...     ...   
6812     \nAdventure, Drama, Fantasy                 5.0   
6813                         \nDrama                 5.0   
6814               \nDrama, Thriller                 5.0   
6815  \nAnimation, Action, Adventure                 5.0   
6816            \nDocumentary, Sport                 5.0   

                                               ONE-LINE  \
0     \nA woman with a mysterious illness is forced ...   
1     \nThe war for Eternia begins again in what may...   
2     \nSheriff Deputy Rick Grimes wakes up from a c...   
3     \nAn animated series that follows the exploits...   
4     \nA prequel, set before the events of Army of ...   
...                                                 ...   
6812                                     \nAdd a Plot\n   
6813                                     \nAdd a Plot\n   
6814                                     \nAdd a Plot\n   
6815                                     \nAdd a Plot\n   
6816                                     \nAdd a Plot\n   

                                                  STARS    VOTES  RunTime  
0     \n    Director:\nPeter Thorwarth\n| \n    Star...   21,062    121.0  
1     \n            \n    Stars:\nChris Wood, \nSara...   17,870     25.0  
2     \n            \n    Stars:\nAndrew Lincoln, \n...  885,805     44.0  
3     \n            \n    Stars:\nJustin Roiland, \n...  414,849     23.0  
4     \n    Director:\nMatthias Schweighöfer\n| \n  ...      NaN      NaN  
...                                                 ...      ...      ...  
6812  \n            \n    Stars:\nMorgan Taylor Camp...      NaN      NaN  
6813  \n    Director:\nTima Shomali\n| \n    Stars:\...      NaN      NaN  
6814  \n    Director:\nNicolai Rohde\n| \n    Stars:...      NaN      NaN  
6815                                                 \n      NaN      NaN  
6816  \n    Director:\nOrlando von Einsiedel\n| \n  ...      NaN      NaN  

[6817 rows x 9 columns]

In [ ]:
def votetoint(x) :
  print(x)
  if Pd.isnull(x):
    return x
  g = [int(i) for i in x.split(sep=",")]
  w=0
  [w:=(w*(10**3))+i for i in g]
  return w
